llvmPackages_20.libc-full: fix building

RossComputerGuy opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

Fixes unpacking so LLVM libc can build.

Fixes Hydra build https://hydra.nixos.org/job/nixpkgs/trunk/llvmPackages_20.libc-full.aarch64-linux

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossComputerGuy committed:
llvmPackages_20.libc-full: fix building

alyssais commented on pkgs/development/compilers/llvm/common/libc/default.nix:
```diff
@@ -46,6 +46,7 @@ stdenv.mkDerivation (finalAttrs: {
   outputs = [ "out" ] ++ (lib.optional isFullBuild "dev");
 
   postUnpack = lib.optionalString isFullBuild ''
+    chmod u+rw $sourceRoot/../$pname
```
This non-recursive chmod does not fix the build, because it does not affect the directory where patchShebangs is trying to add a temporary file.  Perhaps you want this?

```suggestion
    chmod +w $sourceRoot/../$pname/utils/hdrgen
```

Remember to test your changes!

RossComputerGuy replied:
Huh wait why did it work for me then?

RossComputerGuy replied:
Oh, I accidentally did an overlay build. I've fixed full now.

alyssais reviewed (changes requested):
Commit message should start with "llvmPackages_20.libc-full:", because "llvmPackages.libc" doesn't exist and means OfBorg does the wrong thing.

alyssais commented:
Still doesn't seem to build in the default configuration according to OfBorg.  If the CMake compiler working tests are failing something is seriously wrong — overriding them is never going to be the right answer.

But presumably just the chmod should fix building in some pkgsLLVM configuration, so we could go ahead with just that.

RossComputerGuy commented:
> If the CMake compiler working tests are failing something is seriously wrong — overriding them is never going to be the right answer.

Even if it's trying to use a libc when one isn't available?

alyssais commented:
In that case it seems a bit strange for LLVM libc to be having CMake do those checks…

RossComputerGuy commented:
LLVM libc doesn't control that behavior, it comes from CMake itself. I believe we have to enable the same flags to force CMake to not check if the compilers work in compiler-rt if there's no libc. I don't think there's any way to get around this.

alyssais commented:
Hmm okay, I guess if we already do it in compiler-rt…

RossComputerGuy commented on pkgs/development/compilers/llvm/common/libc/default.nix:
```diff
@@ -65,10 +66,22 @@ stdenv.mkDerivation (finalAttrs: {
+      (lib.cmakeBool "CMAKE_CXX_COMPILER_WORKS" true)
+    ];
+
+  NIX_CFLAGS_COMPILE = lib.optionalString isFullBuild "-Wno-unused-command-line-argument";
```
I believe LLVM libc already has that flag set.

RossComputerGuy replied:
No? You were mentioning adding `-nostartfiles` to the CFLAGS.

alyssais replied:
Ah, right. So that sounds like it would be better to not set it in our wrapper and let build systems take care of it, so we don't end up trying passing it in contexts where it'll have no effect and produce an error?

RossComputerGuy replied:
Maybe, I think what we have is fine. This seems to work currently.

alyssais replied:
Well it doesn't exactly work, because we're having to hack around it here.  So we need to discuss what the right solution is.

RossComputerGuy replied:
I'm not sure what the problem is. The solution is to disable the checks CMake runs because it only works with a libc. We cannot build a libc when there's no libc so we just disable the checks. Unless we can build a libc for target and use it in a way that it doesn't cause infinite recursion. This way is just the simplest imo.

alyssais replied:
This thread isn't about the CMake checks?  I already said I'm happy with those.  This is about `-Wno-unused-command-line-argument`, which we only need because we're making our wrapper (that is only used for two packages) pass an argument that doesn't do anything for one of those packages.

RossComputerGuy replied:
Oh, I didn't know it wasn't about the CMake stuff since GitHub shows like 10 lines this thread is about.

RossComputerGuy replied:
Now I see what you mean:
```
libc> /nix/store/h3wgz6n8bc4n61vv427xl8cz69vcd96c-clang-wrapper-20.1.0-rc3/bin/clang++ -DLIBC_NAMESPACE=__llvm_libc_20_1_0_rc3 -I/build/libc-src-20.1.0-rc3/libc -isystem /build/libc-src-20.1.0-rc3/runtimes/build/libc/include -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wno-comment -Wstring-conversion -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG -std=gnu++17 -DLIBC_QSORT_IMPL=LIBC_QSORT_QUICK_SORT -DLIBC_ADD_NULL_CHECKS -fpie -ffreestanding -DLIBC_FULL_BUILD -nostdlibinc -idirafter/usr/include -ffixed-point -fno-builtin -fno-exceptions -fno-lax-vector-conversions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti -ftrivial-auto-var-init=pattern -fno-omit-frame-pointer -Wall -Wextra -Werror -Wconversion -Wno-sign-conversion -Wno-c99-extensions -Wno-gnu-imaginary-constant -Wno-pedantic -Wimplicit-fallthrough -Wwrite-strings -Wextra-semi -Wnewline-eof -Wnonportable-system-include-path -Wstrict-prototypes -Wthread-safety -Wglobal-constructors -DLIBC_COPT_PUBLIC_PACKAGING -MD -MT libc/src/ctype/CMakeFiles/libc.src.ctype.isdigit.dir/isdigit.cpp.o -MF libc/src/ctype/CMakeFiles/libc.src.ctype.isdigit.dir/isdigit.cpp.o.d -o libc/src/ctype/CMakeFiles/libc.src.ctype.isdigit.dir/isdigit.cpp.o -c /build/libc-src-20.1.0-rc3/libc/src/ctype/isdigit.cpp
libc> clang++: error: argument unused during compilation: '-nostartfiles' [-Werror,-Wunused-command-line-argument]
libc> [177/1147] Building CXX object libc/startup/linux/aarch64/CMakeFiles/libc.startup.linux.aarch64.start.dir/start.cpp.o
libc> FAILED: libc/startup/linux/aarch64/CMakeFiles/libc.startup.linux.aarch64.start.dir/start.cpp.o
```

RossComputerGuy replied:
I've dropped the `-nostartfiles` from the CC wrapper in the stdenv being passed to `libc-full` but I've made it as an override so it doesn't potentially break anything else.

alyssais replied:
Correct me if I'm wrong, because I might have got the details wrong since this has been going a while, but it's only used in two places.  There's only one other thing that could potentially break.  We shouldn't need to complicate things with an override.

RossComputerGuy replied:
I'm not sure, I didn't want to introduce any additional rebuilds with this PR.

alyssais replied:
Avoiding one-off rebuilds shouldn't be nearly as big of a priority as keeping our code maintainable forever. Hacks for the former come at the expense of the latter.

RossComputerGuy replied:
Ok, I've dropped the flag there.

RossComputerGuy replied:
And now it seems this will have to go to staging.

alyssais replied:
(I'm happy to help with testing all these as my time allows and build capacity permits btw.)

alyssais replied:
Okay that wasn't actually nearly as much rebuilding as the labels had me believe — looks like everything is still working. If you want to test stuff in pkgsLLVM or Darwin that might be helpful but from my side I'm happy.

alyssais reviewed (changes requested):
The second commit seems to be doing several things at once, and should be split.

RossComputerGuy committed:
llvmPackages_20.libc-full: fix infinite recursion with pkgsLLVMLibc in cmake & python3

RossComputerGuy commented:
> The second commit seems to be doing several things at once, and should be split.

I split it and apparently this time I didn't need everything for `pkgsLLVMLibc.stdenv` to eval.

alyssais commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -932,18 +932,12 @@ let
-            ) "-fno-exceptions";
+          nixSupport.cc-cflags = lib.optional (
+            lib.versionAtLeast metadata.release_version "15" && stdenv.targetPlatform.isWasm
+          ) "-fno-exceptions";
```
Maybe this should be its own commit with its own explanation?  The motivation is that this was added for pkgsLLVM, but doesn't seem to be necessary any more and is causing problems where clang errors on arguments that don't do anything.

alyssais replied:
Still no explanation in the commit message, and I think it needs it.

RossComputerGuy replied:
Commit or comment? I left a comment.

alyssais replied:
Either is fine, as long as it's explained somewhere.

RossComputerGuy replied:
Done

alyssais commented on pkgs/development/compilers/llvm/common/default.nix:
```diff
@@ -1219,6 +1220,14 @@ let
           stdenv = overrideCC stdenv buildLlvmTools.clangNoLibcNoRt;
+
+          cmake =
+            if stdenv.targetPlatform.libc == "llvm" then buildPackages.cmakeMinimal else buildPackages.cmake;
```
Might be nicer to have this on multiple lines too so it matches?

alyssais replied:
Ah didn't realise this file was formatted. No problem then.

RossComputerGuy committed:
llvmPackages.clangNoLibcNoRt: drop nostartfiles flag

`-nostartfiles` used to be needed for pkgsLLVM, causes problems so don't
include it.

RossComputerGuy committed:
lib.systems: mark LLVM libc as static

RossComputerGuy committed:
llvmPackages_20.libc: enable scudo & link Scrt1

RossComputerGuy commented:
I've gone ahead and mark `pkgsLLVMLibc` as static since it doesn't have a dynamic linker, this required symlinking `Scrt1.o` as `crt1.o` since the LLVM libc CMake doesn't do that. I've also added scudo as detailed in https://libc.llvm.org/full_host_build.html so we can get malloc.

alyssais commented on lib/systems/default.nix:
```diff
@@ -200,7 +200,8 @@ let
       # has the `staticMarker`.
-      isStatic = final.isWasi || final.isRedox;
+      isStatic = final.isWasi || final.isRedox ||
+        final.parsed.abi.name == "llvm" || final.libc == "llvm";
```
Should we have an isLLVMLibc?

RossComputerGuy replied:
We do have one, I forgot I added it in.

RossComputerGuy replied:
I've switched over to using it.

alyssais commented on pkgs/development/compilers/llvm/common/libc/default.nix:
```diff
@@ -60,18 +61,26 @@ stdenv.mkDerivation (finalAttrs: {
+      substituteAll ${./libc-shim.tpl} $out/lib/libc.so
+    ''
+    + lib.optionalString (isFullBuild && stdenv.hostPlatform.isLinux) ''
+      ln $out/lib/crt1.o $out/lib/Scrt1.o
```
Do you know why the build system doesn't do this?

RossComputerGuy replied:
Because the libc doesn't recognize when it's compiling for static or dynamic yet. Also doesn't have a dynamic linker.

alyssais replied:
Makes sense — could you note that in a comment so it's clear if it should be changed in future?

RossComputerGuy replied:
Done

alyssais reviewed (approved):
Thanks for sticking with this! I think it's in good shape now.
